File: writers/32bit/cpuidwrt.c
#include <stdio.h>
#include <stdint.h>
#include <cpuid.h>
#include <stdlib.h>

typedef struct {
    uint32_t leaf;
    uint32_t subleaf;
    uint32_t eax;
    uint32_t ebx;
    uint32_t ecx;
    uint32_t edx;
} CPUIDEntry;

#define INITIAL_ENTRIES 64

int main(void) {
    CPUIDEntry *entries = NULL;
    size_t count = 0, capacity = INITIAL_ENTRIES;
    FILE *fp;
    uint32_t maxStandard, maxExtended;
    uint32_t eax, ebx, ecx, edx;
    uint32_t leaf, subleaf;

    entries = malloc(capacity * sizeof(CPUIDEntry));
    if (!entries) {
        fprintf(stderr, "Memory allocation error.\n");
        return 1;
    }

    #define APPEND_ENTRY(LF, SL, A, B, C, D) do {         \
        if(count >= capacity) {                           \
            capacity *= 2;                                \
            entries = realloc(entries, capacity * sizeof(CPUIDEntry)); \
            if(!entries) {                                \
                fprintf(stderr, "Memory reallocation error.\n");       \
                exit(1);                                  \
            }                                             \
        }                                                 \
        entries[count].leaf = (LF);                       \
        entries[count].subleaf = (SL);                    \
        entries[count].eax = (A);                         \
        entries[count].ebx = (B);                         \
        entries[count].ecx = (C);                         \
        entries[count].edx = (D);                         \
        count++;                                          \
    } while(0)

    /* --- Capture Standard CPUID Leaves --- */
    if (!__get_cpuid(0, &maxStandard, &ebx, &ecx, &edx)) {
        fprintf(stderr, "CPUID not supported.\n");
        free(entries);
        return 1;
    }
    for (leaf = 0; leaf <= maxStandard; leaf++) {
        if (leaf == 4 || leaf == 0xB || leaf == 0xD) {
            subleaf = 0;
            while (1) {
                __get_cpuid_count(leaf, subleaf, &eax, &ebx, &ecx, &edx);
                if (leaf == 4 && subleaf > 0 && ((eax & 0x1F) == 0))
                    break;
                if (leaf == 0xB && subleaf > 0 && (eax == 0))
                    break;
                if (leaf == 0xD && subleaf > 0 && (eax==0 && ebx==0 && ecx==0 && edx==0))
                    break;
                APPEND_ENTRY(leaf, subleaf, eax, ebx, ecx, edx);
                subleaf++;
            }
        } else {
            __get_cpuid(leaf, &eax, &ebx, &ecx, &edx);
            APPEND_ENTRY(leaf, 0, eax, ebx, ecx, edx);
        }
    }

    /* --- Capture Extended CPUID Leaves --- */
    __get_cpuid(0x80000000, &maxExtended, &ebx, &ecx, &edx);
    for (leaf = 0x80000000; leaf <= maxExtended; leaf++) {
        if (leaf == 0x8000001D) {
            subleaf = 0;
            while (1) {
                __get_cpuid_count(leaf, subleaf, &eax, &ebx, &ecx, &edx);
                if (subleaf > 0 && ((eax & 0x1F) == 0))
                    break;
                APPEND_ENTRY(leaf, subleaf, eax, ebx, ecx, edx);
                subleaf++;
            }
        } else {
            __get_cpuid(leaf, &eax, &ebx, &ecx, &edx);
            APPEND_ENTRY(leaf, 0, eax, ebx, ecx, edx);
        }
    }

    /* --- Write JSON Output --- */
    fp = fopen("cpuid_data.json", "w");
    if (!fp) {
        fprintf(stderr, "Error opening output file.\n");
        free(entries);
        return 1;
    }

    fprintf(fp, "{\n  \"entries\": [\n");
    for (size_t i = 0; i < count; i++) {
        fprintf(fp,
            "    { \"leaf\": %u, \"subleaf\": %u, \"eax\": %u, \"ebx\": %u, \"ecx\": %u, \"edx\": %u }%s\n",
            entries[i].leaf,
            entries[i].subleaf,
            entries[i].eax,
            entries[i].ebx,
            entries[i].ecx,
            entries[i].edx,
            (i + 1 == count) ? "" : ","
        );
    }
    fprintf(fp, "  ]\n}\n");
    fclose(fp);

    printf("CPUID data captured in cpuid_data.json\n");
    free(entries);
    return 0;
}

File: writers/32bit/cpuidwrt.pas
program cpuidwrt;

{$APPTYPE CONSOLE}
{$MODE DELPHI}  { Delphi mode for better compatibility }
{$IFDEF UNIX}
  {$DEFINE POSIX}
{$ENDIF}

uses
  SysUtils, Classes;

type
  TEntry = record
    Leaf: LongWord;
    SubLeaf: LongWord;
    EAX: LongWord;
    EBX: LongWord;
    ECX: LongWord;
    EDX: LongWord;
  end;

var
  Entries: array of TEntry;

{------------------------------------------------------------
  cpuid executes the CPUID instruction.
  Parameters:
    leaf, subleaf: Input registers.
    eaxOut, ebxOut, ecxOut, edxOut: Output registers.
  Note: We push/pop EBX because it is callee‐saved.
-------------------------------------------------------------}
procedure cpuid(Leaf, SubLeaf: LongWord; var EAXOut, EBXOut, ECXOut, EDXOut: LongWord);
asm
  push ebx
  mov eax, Leaf
  mov ecx, SubLeaf
  cpuid
  mov EBXOut, ebx
  mov EAXOut, eax
  mov ECXOut, ecx
  mov EDXOut, edx
  pop ebx
end;

procedure AppendEntry(Leaf, SubLeaf, EAX, EBX, ECX, EDX: LongWord);
var
  L: Integer;
begin
  L := Length(Entries);
  SetLength(Entries, L + 1);
  with Entries[L] do
  begin
    Self.Leaf := Leaf;
    Self.SubLeaf := SubLeaf;
    Self.EAX := EAX;
    Self.EBX := EBX;
    Self.ECX := ECX;
    Self.EDX := EDX;
  end;
end;

procedure CaptureData;
var
  leaf, subleaf: LongWord;
  maxStandard, maxExtended: LongWord;
  a, b, c, d: LongWord;
begin
  SetLength(Entries, 0);

  {--- Capture Standard CPUID Leaves ---}
  cpuid(0, 0, maxStandard, b, c, d);  { maxStandard is in EAX }
  for leaf := 0 to maxStandard do
  begin
    if (leaf = 4) or (leaf = $B) or (leaf = $D) then
    begin
      subleaf := 0;
      while True do
      begin
        cpuid(leaf, subleaf, a, b, c, d);
        if (leaf = 4) and (subleaf > 0) and ((a and $1F) = 0) then Break;
        if (leaf = $B) and (subleaf > 0) and (a = 0) then Break;
        if (leaf = $D) and (subleaf > 0) and (a = 0) and (b = 0) and (c = 0) and (d = 0) then Break;
        AppendEntry(leaf, subleaf, a, b, c, d);
        Inc(subleaf);
      end;
    end
    else
    begin
      cpuid(leaf, 0, a, b, c, d);
      AppendEntry(leaf, 0, a, b, c, d);
    end;
  end;

  {--- Capture Extended CPUID Leaves ---}
  cpuid($80000000, 0, maxExtended, b, c, d);  { maxExtended in EAX }
  for leaf := $80000000 to maxExtended do
  begin
    if (leaf = $8000001D) then
    begin
      subleaf := 0;
      while True do
      begin
        cpuid(leaf, subleaf, a, b, c, d);
        if (subleaf > 0) and ((a and $1F) = 0) then Break;
        AppendEntry(leaf, subleaf, a, b, c, d);
        Inc(subleaf);
      end;
    end
    else
    begin
      cpuid(leaf, 0, a, b, c, d);
      AppendEntry(leaf, 0, a, b, c, d);
    end;
  end;
end;

procedure WriteJSON(const AFileName: string);
var
  i: Integer;
  SL: TStringList;
begin
  SL := TStringList.Create;
  try
    SL.Add('{');
    SL.Add('  "entries": [');
    for i := 0 to High(Entries) do
    begin
      with Entries[i] do
      begin
        { For simplicity, every entry is followed by a comma except the last }
        if i < High(Entries) then
          SL.Add(Format('    { "leaf": %u, "subleaf": %u, "eax": %u, "ebx": %u, "ecx": %u, "edx": %u },',
            [Leaf, SubLeaf, EAX, EBX, ECX, EDX]))
        else
          SL.Add(Format('    { "leaf": %u, "subleaf": %u, "eax": %u, "ebx": %u, "ecx": %u, "edx": %u }',
            [Leaf, SubLeaf, EAX, EBX, ECX, EDX]));
      end;
    end;
    SL.Add('  ]');
    SL.Add('}');
    SL.SaveToFile(AFileName);
  finally
    SL.Free;
  end;
end;

begin
  try
    CaptureData;
    WriteJSON('cpuid_data.json');
    Writeln('CPUID data captured in cpuid_data.json');
  except
    on E: Exception do
      Writeln('Error: ', E.Message);
  end;
end.
